remotes/docker/pusher.go

```diff
@@ -393,9 +417,11 @@ func (pw *pushWriter) Write(p []byte) (n int, err error) {
 		// if the pipe is closed, we might have the original error on the error
 		// channel - so we should try and get it
 		select {
-		case err2 := <-pw.errC:
-			err = err2
-		default:
+		case <-pw.done:
```

Comment:
Let me make sure I understand the 3 cases correctly...
> - The request has failed, we need to attempt a reset, so we can expect a
  new pipe incoming on pipeC.

A redirect or something that retries where the http library requests a new body

> - The request has failed, we don't need to attempt a reset, so we can
  expect an incoming error on errC.

Either the request fails or an unexpected status code is returned

> - Something else externally has called Close, so we can expect the done
  channel to be closed.

straightforward, but don't see any case where this happens

There is a condition where the registry (unexpectedly, not to spec) returns 201 or 204 on the put before the body is fully written. I would expect that the http library would issue close and could fall into a deadlock here. We could just read `respC` and call `setResponse`.  In that case `ErrClosedPipe` would get returned and `Commit` shouldn't be called anyway.